Feature/core batch search endpoint

aristotelisxs opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Extending the core_search function's capabilities to include multiple search queries in case results need to be retrieved from a wide range of requests in a more efficient manner.

## Description
<!--- Describe your changes in detail -->
Adding in a POST request to the core_search function and a convenience function to the zzz.R file that will convert a simple character vector, e.g. c("data mining", "machine learning", "semantic web"), to a suitable JSON object required from the endpoint to perform a batch search request. This includes updating of the documentation with the added parameter to the function (called "method" for consistency with the similar core_articles function's parameter) and checks to disallow multiple queries in GET requests for the core_search function.

Moreover, single valued parameters in the JSON objects sent to the endpoints will be unboxed as this is a requirement for all of CORE's endpoints.

## Example
<!--- if introducing a new feature or changing behavior of existing
methods/functions, include an example if possible to do in brief form -->
As included in the documentation:

query <- c("data mining", "machine learning", "semantic web")
res <- core_search(query, method = "POST")
head(res$data)
res$data$id

The "page" and "pageSize" variables will be constant in all individual query requests (may not be the desirable functionality for some but done so for the sake of simplicity).

aristotelisxs committed:
Extends the search api to post batch requests on multiple article ids. This includes a convenience function that, when supplied a list of queries, will construct a nested list for use in the request.

aristotelisxs committed:
Unbox elements in JSON elements of POST requests (this is a requirement for the batch search operation). Fill in further arguments for core_POST on batch search.

aristotelisxs committed:
Extra check on the core_search GET method to disallow character vectors (only allowed on the new POST request). Updated the core_search markdown with an extra example and the addition of the extra parameter (“method”) of the updated function.

aristotelisxs committed:
Added missing “method” parameter method on the markdown’s function declarations.

aristotelisxs committed:
Missed out the method parameters on the markdown.

aristotelisxs committed:
Convenience function should list pageSize as the limit parameter instead.

sckott commented:
thanks @aresxs91 

First, a question. Would it be easier to instead of having a new parameter `method` to use `POST` when `length(query) > 1` and `GET` otherwise ?

aresxs91 commented:
Hello Scott,

Thanks for your comment. I see, yes that would indeed make it more concise.
I copied your style for the core_articles function there to be honest. I'll
revise it and let you know :)

Best,
Aris

On Wed, 19 Apr 2017 at 20:10 Scott Chamberlain <notifications@github.com>
wrote:

> thanks @aresxs91 <https://github.com/aresxs91>
>
> First, a question. Would it be easier to instead of having a new parameter
> method to use POST when length(query) > 1 and GET otherwise ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ropenscilabs/cored/pull/3#issuecomment-295396302>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKcM14JrLqVDzYl0e43SgqJf5ogJNvWzks5rxlwzgaJpZM4NBy9S>
> .
>

aristotelisxs committed:
Removing the requirement for a “method” parameter on the core_search function, replaced by a check on the size of queries to process on the search endpoint (if larger than 1, a POST request will be automatically issued). Cleanup of comments and correction of the function’s documentation accordingly.

aristotelisxs committed:
Removing validity checks for page and pageSize on the batch query list creation function as they are dealt with further up the hierarchy.

aristotelisxs committed:
Documentation link to searchBatch on CORE.

sckott commented:
@aresxs91 ping me when you're done

aresxs91 commented:
It seems alright from my end, I haven't made extensive tests for the function but it looks ready.

sckott commented on R/core_search.R:
```diff
@@ -28,6 +28,17 @@ core_search <- function(query, page = 1, limit = 10, key = NULL,
   must_be(limit)
-  core_GET(path = file.path("search", query), key,
-           list(page = page, pageSize = limit), ...)
+  switch(
```
since there's only two options, an if/else makes more sense here i think

aristotelisxs replied:
Agreed.

sckott commented on R/zzz.R:
```diff
@@ -25,7 +25,7 @@ core_POST <- function(path, key, args, body, ...){
     path = file.path("api-v2", path),
     query = cp(args),
-    body = jsonlite::toJSON(body), encode = "json", ...
+    body = jsonlite::toJSON(body, auto_unbox=T), encode = "json", ...
```
please use `TRUE` and `FALSE` instead of `T` and `F`

sckott commented on R/zzz.R:
```diff
@@ -74,3 +74,14 @@ clog <- function(x){
+
+create_batch_query_list <- function(queries, page, pageSize) {
+  queryList <- lapply(queries, function(x){
+    as.list(setNames(x, rep("query", length(x))))
```
namespace `setNames` here like `stats::setNames` since we don't import the stats pkg

sckott reviewed (changes requested):
just a few changes

aristotelisxs committed:
Changes include:
1) If/else instead of switch because of only two possible outcomes.
2) TRUE instead of T within functions.
3) Specifying namespace “stats” for function “setNames”

sckott commented:
thanks very much!
